### ISSUE TITLE
Deletion of githubsource does not always delete webhook

### ISSUE THREAD

**Issue by dibbles** [knative/eventing-contrib#216]
## Expected Behavior

If I delete a githubsource the associated webhook should be deleted

## Actual Behavior

Sometimes deletion occurs, but more often than not deletion does not occur.

## Steps to Reproduce the Problem

  1. kubectl apply -f <githubsource.yaml>
  2. check in github repo webhooks to see webhook created
  3. kubectl delete -f <githubsource.yaml>
  4. check in github repo webhooks to see if webhook created or not
  5. repeat a number of times and you will just see an accumulation of webhooks 

## Additional Info

**Comment by a-roberts** [comment]
This is definitely a problem and I see it 1/3 of the time. 

The important pod that does webhook reconciliation, or at least reports on it (e.g. for a create or delete on GitHub ) is `controller-manager-0` sitting in `knative-sources`. 

For ease of reading (omitting line numbers and other logging info), here are the logs.

Good delete:

```
"msg":"Reconciling &TypeMeta{Kind:,APIVersion:,} default/jess88"}
"msg":"deleting GitHub webhook"}
"msg":"delete webhook error response:\n&{Response:0xc000d7c000 NextPage:0 PrevPage:0 FirstPage:0 LastPage:0 Rate:github.Rate{Limit:5000, Remaining:5000, Reset:github.Timestamp{2019-06-06 13:44:09 +0000 UTC}}}"}
"msg":"Failed to reconcile &TypeMeta{Kind:,APIVersion:,}: failed to delete webhook: failed to delete the webhook: DELETE https://github.ibm.com/api/v3/repos/AROBERTS/icp-nodejs-sample/hooks/5141345: 404 Not Found []"}
"msg":"Failed to update","error":"GitHubSource.sources.eventing.knative.dev \"jess88\" not found","objectKind":"&TypeMeta{Kind:,APIVersion:,}","stacktrace":"github.com/knative/eventing-sources/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/go/src/github.com/knative/eventing-sources/pkg/controller/sdk/reconciler.go:83\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
"msg":"Reconciling &TypeMeta{Kind:,APIVersion:,} default/jess89"}
"msg":"deleting GitHub webhook"}
```

Bad delete:

```
"msg":"Reconciling &TypeMeta{Kind:,APIVersion:,} default/jess90"}
"msg":"Failed to update","error":"GitHubSource.sources.eventing.knative.dev \"jess90\" not found","objectKind":"&TypeMeta{Kind:,APIVersion:,}","stacktrace":"github.com/knative/eventing-sources/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/go/src/github.com/knative/eventing-sources/pkg/controller/sdk/reconciler.go:83\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
"msg":"Reconciling &TypeMeta{Kind:,APIVersion:,} default/jess90"}
"msg":"could not find &TypeMeta{Kind:,APIVersion:,} default/jess90\n","stacktrace":"github.com/knative/eventing-sources/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/go/src/github.com/knative/eventing-sources/pkg/controller/sdk/reconciler.go:59\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
"msg":"Reconciling &TypeMeta{Kind:,APIVersion:,} default/jess90"}
"msg":"could not find &TypeMeta{Kind:,APIVersion:,} default/jess90\n","stacktrace":"github.com/knative/eventing-sources/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/go/src/github.com/knative/eventing-sources/pkg/controller/sdk/reconciler.go:59\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/knative/eventing-sources/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

**Notice the not found error.** 

**My guess is, we're not waiting for the webhook to be deleted (perhaps as part of a "finaliser" when the GitHubSource is going away)**.

Important delete code I've found is as follows.

https://github.com/knative/eventing-sources/blob/b6cdf390669a28bc0d6808476851e2a34aa94305/pkg/reconciler/githubsource/githubsource.go#L266

and this is using `gitHubWebhookClient` which has a delete method here:

https://github.com/knative/eventing-sources/blob/6c45382e71edbd2096a26124df7be2f290473f75/pkg/reconciler/githubsource/webhook_client.go#L79

This uses `ghClient`, from Google: `ghclient "github.com/google/go-github/github"` and that code is here:

https://github.com/google/go-github/blob/d173fe318e90d040e1c08ced393dc7afabaadea0/github/repos_hooks.go#L195

**TL;DR**: I don't see any *wait around and verify it's really gone* checks, or attempts being made, so guessing there's a race condition here and hence the lack of it happening all the time. Perhaps we should allow up to a 500ms wait and poll as this client doesn't always work.

**Comment by duglin** [comment]
I'm seeing this. For me it never gets deleted and is kind of a big issue since deleting webhook on github is annoying since you have to click twice to delete each one. So, if you're running some tests you can get a ton of them.

/cc @daisy-ycguo

**Comment by duglin** [comment]
not 100% sure it's related but when I deleted the event source I noticed these in the github controller logs - it might shed some light on things:

```
{
  "level": "error",
  "ts": "2019-08-13T02:24:49.849Z",
  "caller": "sdk/reconciler.go:84",
  "msg": "Failed to update",
  "knative.dev/controller": "github-controller-manager",
  "request": "default/githubsource",
  "error": "Operation cannot be fulfilled on githubsources.sources.eventing.knative.dev \"githubsource\": StorageError: invalid object, Code: 4, Key: /registry/sources.eventing.knative.dev/githubsources/default/githubsource, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 34168df8-bd61-11e9-8ec6-f2c1ba59b1e5, UID in object meta: ",
  "objectKind": "&TypeMeta{Kind:,APIVersion:,}",
  "stacktrace": "github.com/knative/eventing-contrib/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/root/go/src/github.com/knative/eventing-contrib/pkg/controller/sdk/reconciler.go:84\ngithub.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
}
```
and 
```
{
  "level": "error",
  "ts": "2019-08-13T02:24:50.849Z",
  "caller": "sdk/reconciler.go:60",
  "msg": "could not find &TypeMeta{Kind:,APIVersion:,} default/githubsource\n",
  "knative.dev/controller": "github-controller-manager",
  "request": "default/githubsource",
  "stacktrace": "github.com/knative/eventing-contrib/pkg/controller/sdk.(*Reconciler).Reconcile\n\t/root/go/src/github.com/knative/eventing-contrib/pkg/controller/sdk/reconciler.go:60\ngithub.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/root/go/src/github.com/knative/eventing-contrib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
}
```

**Comment by n3wscott** [comment]
gcppubsubsource has been removed from this repo. Please see http://github.com/google/knative-gcp 

`PullSubscription` is the replacement. 

/close

**Comment by dibbles** [comment]
@n3wscott can you give more details about why you closed this?

- knative/eventing-contrib still exists and has a githubsource sample
- you give no details about what the problem was caused by
- the new repo you reference is marked experimental
- the new repo you reference mentions GCP and knative, is this repo/function specific to GCP?
